### DISCUSSION TITLE
Best practice for mapped React components each with their own store?

### DISCUSSION THREAD

**Discussion by rubenesque-code** [pmndrs/zustand#1682] [Q&A]
Hi there,

I'm using React and Typescript. I'm mapping over an array of components, each of which need their own state. In this case, the state is simple and is about the visibility state of a modal, only one of which can be open at a time. I'm wandering:

1. if it's recommended to use zustand in this case (multiple instances of a store) or if zustand should be seen as a like-for-like replacement for redux.
2. if it is recommended, how do i do it? 
    - From my reading I have to create multiple instances of the store and it seems as if I should use `createStore` and `useStore` along with React's context hooks, but in the docs the only reference I can find is a non-typescript example for migrating.

Apologies if I've missed something glaring in the docs.

**Comment by dai-shi** [comment]
> I should use createStore and useStore along with React's context hooks

Yes, and probably with useRef or `use-constant`.

For TS typing, someone would be able to help if you provide a small reproduction.

**Reply by rubenesque-code** [reply]
Thanks for getting back to me! Maybe the below is okay, but I wasn't sure, particularly with what type to pass to `createContext`  (`ModalVisibilityState` or `typeof store`) and the typing of the selector parameter of  `useModalVisibilityStore`.

```
import { createContext, type ReactElement, useContext } from "react";
import { createStore, useStore } from "zustand";

type ModalVisibilityState = {
  isOpen: boolean;
  openModal: () => void;
  closeModal: () => void;
};

const store = createStore<ModalVisibilityState>()((set) => ({
  isOpen: false,

  openModal: () => set(() => ({ isOpen: true })),

  closeModal: () => set(() => ({ isOpen: false })),
}));

const MyContext = createContext<typeof store>({} as typeof store);

export const ModalVisibilityProvider = ({
  children,
}: {
  children: ReactElement;
}) => <MyContext.Provider value={store}>{children}</MyContext.Provider>;

export const useModalVisibilityStore = (
  selector: (state: ModalVisibilityState) => ModalVisibilityState
) => useStore(useContext(MyContext), selector);
```

**Reply by dai-shi** [reply]
Does this help? https://tsplay.dev/m0Y0rW

Oh, you want to check the existence of provider? You should put `null` as default value then.

**Reply by dai-shi** [reply]
Like this: https://tsplay.dev/wE7qbm

**Reply by rubenesque-code** [reply]
> Like this: https://tsplay.dev/wE7qbm

That's great. Guess I wasn't far off. Not sure why I started asserting the existence of the provider when using `createContext` but I think there's a reason. Now I look at it again I do think I've seen a similar TS example somewhere but was maybe confused by it; the use of the generic in the use hook makes total sense, I just haven't used a generic with a function return type before.

Thanks for your help!

**Comment by antonisprovidakis** [comment]
@dai-shi 

> Yes, and probably with useRef or `use-constant`.


Is something wrong with 
```ts
const [store] = useState(() => createSomeStore())`
```
? 🤔 

Any disadvantages compared to the `useRef` alternative?

**Reply by dai-shi** [reply]
It works, but it comes with an extra tiny cost. useState is basically for a state that changes over time.
